### ISSUE TITLE
Add latest as pre-release on PyPI

### ISSUE THREAD

**Issue by chamberlain2007** [postlund/pyatv#181]
Would it be possible to get the latest packaged as a pre-release version on PyPI? I would like to modify a custom component for HASS to use the MRP code but it requires the package to be installed from PyPI, and only the 0.3.12 version is there.

**Comment by jimz011** [comment]
Seems like it has been updated

https://pypi.org/project/pyatv/0.3.13/

**Comment by postlund** [comment]
Following this up now... Yes, I can try to help you with that. It is much better if you create PRs against this repo with changes to pyatv instead of copying the code. It's a step backwards otherwise (especially as pyatv isn't my main priority).

I can help out with some coaching regarding with what has to be done if you like and try to do code review as fast as I can, assuming you wanna help out getting the Apple TV support on track again?

**Comment by chamberlain2007** [comment]
@jimz011 It doesn't look like 0.3.13 includes the MRP code that's on master.

@postlund I appreciate all your work. I'm happy to work with you to get the work I've done into this repo, I just did what I needed to do to get things up and running for myself. Most of the changes were either to the homeassistant side, or to the pyatv side to get around the fact that the version in pypi didn't have the MRP code. If I could just get the MRP code that's on pyatv master into pypi in a prerelease version, I can get to work properly updating the actual homeassistant component to reference that, and then PR homeassistant itself instead of a custom component.

**Comment by postlund** [comment]
@chamberlain2007 Great! To start, there are a bunch of things that must be sorted out in pyatv as lots of functionality (and tests) are missing for MRP. I would however say that this is the second priority. Let me tell you why...

In DMAP, everything is quite simple. Get your login ID and IP and you are golden. This however doesn't work with MRP as the port used for communication is randomized for each boot of the Apple TV. Or, in theory, it could change at any time. So it is simply not possible to use the same kind of setup (hardcode IP, port, etc. as you have done in your custom component) as the configuration will break once you restart your device. All details must be discovered (according to Apple specification) via zeroconf/bonjour. This is done with `scan_for_apple_tvs`. It is flaky and that should be investigated, but it must be done this way.

This means that devices (especially MRP-devices) must be discovered using `scan_for_apple_tvs` method in Home Assistant. We should force the same concept for both devices, i.e. DMAP and MRP. For this to work, devices found must be uniquely identified. There is a `device_id` available for this, but it isn't truly unique as the IP-address is used when generating the ID (it isn't implemented for MRP). For DMAP I haven't found any unique identifier, so a simple approach might be to use the MAC-address as input when generating the identifier. Would work for MRP as well, but I think we can find something better there. By implementing unique identifiers we can utilize entity registry and use config flows to setup devices. This is the *only* way to create a truly working implementation. Supporting manual configuration can still be added, but that is more of a "developer" feature than a solution.

One more thing that kinda forces the use of a unique identifier is to support the reboot scenario mentioned earlier. Since, if a device rebooted and changed the MRP port, the already setup entity in Home Assistant will not work anymore. So we will have to re-scan for devices in order to establish a new connection and if we cannot pin-point a particular search result to a specific entity, then we cannot restore the connection. Encryption keys, identifiers and such are to be considered static but IP-address and port are dynamic and can be anything.

So, basically what must be done:

1. Fix `device_id` to generate unique identifiers for both DMAP and MRP
2. Update the component to use the entity registry and config flows, so new devices can be found and added via the Integration page (multiple steps, but easiest done at once I guess)
3. Fix remaining things in pyatv

You of course might have to make adjustments in pyatv along the way to make it fit. But mostly it should work to use the API as-is.

**Comment by SalkinDK** [comment]
@postlund @chamberlain2007 Anything us having no great coding skills, can do to help?
The WAF is rambling down since the light doesnt "follow" the tv anymore.
Would love to help, but dont possess great coding skills.

**Comment by chamberlain2007** [comment]
@postlund Thanks for that. Will definitely look at that as time allows. I also have a day job, so time is constrained for me as well.

@SalkinDK I appreciate that. Unfortunately I only have so many hours and energy in the day :)

**Comment by postlund** [comment]
@SalkinDK I think we are in a tricky state at the moment and doing development work is mainly what needs to be done I'm afraid. But testers will be needed at some point.

@chamberlain2007 Great, same thing here (and I have a baby dropping probably next week or so). Time is limited. However, I have made some progress that you might have seen. Pairing works again for instance. I've been thinking a bit about the identifier part, which is not trivial. At first I made the assumption that I could use information from zeroconf to derive a unique identifier, but the services published for DMAP/Home Sharing doesn't contain any really good candidates for that (AirPlay and MediaRemote has several). They don't have a common identifier amongst themselves either, so if we choose the identifier based on a particular service (e.g. if MRP is available then we pick that), then we will have problems if scanning fails to find that service but find other services on the same device. We would end up with multiple identifiers for the same device, which is unfortunate.

The approach I'm leaning on is to use the MAC-address (as mentioned above). But getting the MAC-address of an IP in python is tricky. Especially to keep platform independence. So a 3rd party library, like [getmac](https://github.com/GhostofGoes/getmac) would probably be useful. However, I would like to use as few external libraries as possible, so I'm torn here. But there's probably not any other feasible alternatives.

**Comment by postlund** [comment]
I just pushed #187, which uses the MAC-address as device identifier. It depends on the third party library getmac, which I ran into a small bug with. Reported it at 
GhostofGoes/getmac#50. Will probably wait for acknowledgment on this before merging.

**Comment by Thalagyrt** [comment]
Using the MAC as an identifier won't work in segregated environments FWIW, you'll always get the MAC of whatever device on the same layer 2 network (the firewall between vlans, in this case) sent the packet. My Apple TVs are in a different VLAN than homeassistant, but thankfully I only control one from homeassistant so this shouldn't be a problem for this particular use case. It'll definitely be a problem in some cases though. I'd sooner go off of just the IP and put a big warning that you must statically assign your apple TVs for this to work properly, or at least have the option to work that way instead of using a network identifier that's not guaranteed to be unique as far as homeassistant can see.

**Comment by postlund** [comment]
Ok, I get what you are saying and the MAC-address is per se not the best identifier as it does not uniquely identify the exact device in every different use cases. What I can say is that the identifier used must be unique as it will break the entity registry for users otherwise. Somehow using the IP-address in one way or another is no better than using MAC-address: it just moves the target around. I'm not going to add special workarounds either. I want one solution and I think I have found a decent enough solution for that.

I am going to use the MAC-address and also change the nomenclature to that as well. You see, the problem here isn't that the MAC address isn't unique. It really is. Apple would not issue two devices with the same MAC-address. The problem here is resolution of it. Using ARP obviously does not work, you have a pretty long explanation of that. But, here is the good thing: we can get it via zeroconf. Both MRP and AirPlay announce it ("macAddress" for AirPlay and "deviceid" for AirPlay). I actually saw that they have their own unique identifiers as well, that could be used too, but the MAC-address is simpler.

Since devices that does not support MRP (Apple TV 3 and older) do support AirPlay, I can just snatch the MAC from there. It implies that it's not enough to just find the DMAP-service, but I'm ok with that. Scanning to initially find the device will be mandatory going forward since MRP actually requires it and I do not want different APIs depending on protocol. So I think this is the way I will go.

**Comment by Thalagyrt** [comment]
Oh awesome, if you can get it via ZeroConf across network boundaries that's perfect, and I absolutely agree it's a much better identifier than an IP address. I was purely thinking about a fallback in the event that getting a better identifier wasn't possible.

What I meant about non-uniqueness was not about the devices themselves, but was that traversing into a different physical network layer, you'll only see the first hop router's MAC address if you're inspecting packets to grab the physical address, but that's not the method that was being used here so is not actually relevant, and I kind of forgot to remove that bit when editing. Without knowing that you could grab it via ZeroConf I simply tossed out the first suggestion that came to my sleepy brain. :)

Thanks for getting back to me and looking at this!

**Comment by postlund** [comment]
> Oh awesome, if you can get it via ZeroConf across network boundaries that's perfect, and I absolutely agree it's a much better identifier than an IP address. I was purely thinking about a fallback in the event that getting a better identifier wasn't possible.
> 
> What I meant about non-uniqueness was not about the devices themselves, but was that traversing into a different physical network layer, you'll only see the first hop router's MAC address if you're inspecting packets to grab the physical address, but that's not the method that was being used here so is not actually relevant, and I kind of forgot to remove that bit when editing. Without knowing that you could grab it via ZeroConf I simply tossed out the first suggestion that came to my sleepy brain. :)
> 
> Thanks for getting back to me and looking at this!

Great, then we are on the same page 😄 Your feedback is good and very valid. I wasn't very satisfied with current solution, so it needed a revision anyway. The tricky part will be service discovery in Home Assistant later. Will have to think about that.

**Comment by postlund** [comment]
Also, closing this issue since I have pushed a pre-release already.

**Comment by postlund** [comment]
I did some digging and hooked up my old ATV3 I had laying around. Looks like the old services (`_touch-able._tcp` and `_appletv-v2._tcp`) contain a sweet identifier of some sort in the service name. I tried googling it but I can't find any good reference of what it means, but I suspect that it is unique enough to identify a device. At least a legitimate Apple-device (forked-daapd uses database id for this, so there might be discrepancies in 3rd party software). So I might use that instead. When it comes to MRP, it already has a unique identifier in the zeroconf service that I can use.
